File: python/assistant-stream/src/assistant_stream/assistant_stream_chunk.py
from dataclasses import dataclass
from typing import Any, Union


# Define the data classes for different chunk types
@dataclass
class TextDeltaChunk:
    text_delta: str
    type: str = "text-delta"


@dataclass
class ToolCallBeginChunk:
    tool_call_id: str
    tool_name: str
    type: str = "tool-call-begin"


@dataclass
class ToolCallDeltaChunk:
    tool_call_id: str
    args_text_delta: str
    type: str = "tool-call-delta"


@dataclass
class ToolResultChunk:
    tool_call_id: str
    result: Any
    type: str = "tool-result"


@dataclass
class DataChunk:
    data: Any
    type: str = "data"

@dataclass
class ErrorChunk:
    error: str
    type: str = "error"


# Define the union type for AssistantStreamChunk
AssistantStreamChunk = Union[
    TextDeltaChunk, ToolCallBeginChunk, ToolCallDeltaChunk, ToolResultChunk, DataChunk, ErrorChunk
]

File: packages/assistant-stream/src/core/AssistantStreamChunk.ts
import { ReadonlyJSONValue } from "./utils/json/json-value";

export type PartInit =
  | {
      readonly type: "text" | "reasoning";
    }
  | {
      readonly type: "tool-call";
      readonly toolCallId: string;
      readonly toolName: string;
    }
  | {
      readonly type: "source";
      readonly sourceType: "url";
      readonly id: string;
      readonly url: string;
      readonly title?: string;
    };

export type AssistantStreamChunk = { readonly path: readonly number[] } & (
  | {
      readonly type: "part-start";
      readonly part: PartInit;
    }
  | {
      readonly type: "part-finish";
    }
  | {
      readonly type: "tool-call-args-text-finish";
    }
  | {
      readonly type: "text-delta";
      readonly textDelta: string;
    }
  | {
      readonly type: "annotations";
      readonly annotations: ReadonlyJSONValue[];
    }
  | {
      readonly type: "data";
      readonly data: ReadonlyJSONValue[];
    }
  | {
      readonly type: "step-start";
      readonly messageId: string;
    }
  | {
      readonly type: "step-finish";
      readonly finishReason:
        | "stop"
        | "length"
        | "content-filter"
        | "tool-calls"
        | "error"
        | "other"
        | "unknown";
      readonly usage: {
        readonly promptTokens: number;
        readonly completionTokens: number;
      };
      readonly isContinued: boolean;
    }
  | {
      readonly type: "message-finish";
      readonly finishReason:
        | "stop"
        | "length"
        | "content-filter"
        | "tool-calls"
        | "error"
        | "other"
        | "unknown";
      readonly usage: {
        readonly promptTokens: number;
        readonly completionTokens: number;
      };
    }
  | {
      readonly type: "result";
      readonly result: any;
      readonly isError?: boolean;
    }
  | {
      readonly type: "error";
      readonly error: string;
    }
);
